client/protoclient.go

```diff
@@ -0,0 +1,491 @@
+	}
+
+	return inputName, outputName, nil
+}
+
+// get recursivily all protos needed in a Unmarshal json
+func getKeys(info map[string]interface{}, keysSet map[string]bool) {
+	for k, v := range info {
+		if strings.Contains(k, "*proto") {
+			kew := strings.Replace(k, "*", "", 1)
+			keysSet[kew] = true
+		}
+
+		listofouts, ok := v.([]interface{})
+		if ok {
+			for i := range listofouts {
+				aux, ok := listofouts[i].(map[string]interface{})
+				if !ok {
+					continue
+				}
+				getKeys(aux, keysSet)
+			}
+		}
+
+		aux, ok := v.(map[string]interface{})
+		if !ok {
+			continue
+		}
+		getKeys(aux, keysSet)
+	}
+}
+
+func (pc *ProtoClient) getDescriptors(data string) error {
+	d := []byte(data)
+	var jsonmap interface{}
+	if err := json.Unmarshal(d, &jsonmap); err != nil {
+		return err
+	}
+	m := jsonmap.(map[string]interface{})
+	keysSet := make(map[string]bool)
+	getKeys(m, keysSet)
+
+	// load predefined protos
+	for _, commands := range pc.info.Commands {
+		if commands.input != "" {
+			keysSet[commands.input] = true
+		}
+		if commands.output != "" {
+			keysSet[commands.output] = true
+		}
+	}
+
+	// build commands reference
+	handlers := m["handlers"].(map[string]interface{})
+	for k, v := range handlers {
+		cmdInfo := v.(map[string]interface{})
+		in, out, err := getOutputInputNames(cmdInfo)
+		if err != nil {
+			return err
+		}
+
+		var command Command
+		command.input = in
+		command.output = out
+
+		pc.info.Commands[k] = &command
+		if pc.descriptorsRoute == "" && in == "protos.ProtoName" && out == "protos.ProtoDescriptor" {
+			pc.descriptorsRoute = k
+		}
+		// c.Println(k);
+	}
+
+	remotes := m["remotes"].(map[string]interface{})
+	for k, v := range remotes {
+		cmdInfo := v.(map[string]interface{})
+		in, out, err := getOutputInputNames(cmdInfo)
+		if err != nil {
+			return err
+		}
+
+		var command Command
+		command.input = in
+		command.output = out
+
+		pc.info.Commands[k] = &command
+		// c.Println(k);
+	}
+
+	// get all proto types
+	descriptorArray := make([]*protobuf.FileDescriptorProto, 0)
+
+	for key := range keysSet {
+		protname := &protos.ProtoName{
+			Name: key,
+		}
+		data, err := proto.Marshal(protname)
+		if err != nil {
+			return err
+		}
+		_, err = pc.SendRequest(pc.descriptorsRoute, data)
+		if err != nil {
+			return err
+		}
+
+		response := <-pc.Client.IncomingMsgChan
+
+		protodecrip := &protos.ProtoDescriptor{}
+
+		if err := proto.Unmarshal(response.Data, protodecrip); err != nil {
+			return err
+		}
+
+		fileDescriptorProto, err := unpackDescriptor(protodecrip.Desc)
+		if err != nil {
+			return err
+		}
+
+		descriptorArray = append(descriptorArray, fileDescriptorProto)
+		pc.descriptorsNames[key] = true
+	}
+
+	err := pc.buildProtosFromDescriptor(descriptorArray)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func new(docslogLevel logrus.Level, requestTimeout ...time.Duration) *ProtoClient {
+	return &ProtoClient{
+		Client:           *New(docslogLevel, requestTimeout...),
```

Comment:
New already returns a pointer

Reply:
ProtoClient extends the Client struct. Here I get the value of the pointer returned by New.